frontend/src/Component/Pages/Homepage.jsx

```diff
@@ -17,44 +18,70 @@ export const Homepage = () => {
         const response = await fetch(`${apiKey}/api/v1/devices`);
         const data = await response.json();
         // Getting the data from the api fetch for room number and dev_eui
-        const extractedData = data.data.map(device => ({
+        const extractedData = data.data.map((device) => ({
           room_number: device.room_number,
-          dev_eui: device.dev_eui,
+          dev_eui: device.dev_eui
         }));
         setDevices(extractedData);
 
         const co2Data = {};
-        await Promise.all(extractedData.map(async (device) => {
-          // This fetches the co2 level for each room
-          const co2Response = await fetch(`${apiKey}/api/v1/rooms/latest/${device.dev_eui}`);
-          const co2Info = await co2Response.json();
-          co2Data[device.dev_eui] = co2Info.data.co2;
-        }));
+        await Promise.all(
+          extractedData.map(async (device) => {
+            // This fetches the co2 level for each room
+            const co2Response = await fetch(
+              `${apiKey}/api/v1/rooms/latest/${device.dev_eui}`
+            );
+            const co2Info = await co2Response.json();
+            co2Data[device.dev_eui] = co2Info.data.co2;
+          })
+        );
         setCo2Levels(co2Data);
         setTemperatures(tempData);
+        setIsLoading(false);
       } catch (error) {
         console.error('Error fetching devices or CO2 levels:', error);
+        setIsLoading(false);
```

Comment:
Instead of setIsLoading at the end of the try block and in the catch block

use a finally after the catch

```js
catch (error) {
... code
} finally {
setIsLoading(false);
}
```

[This reduces any kind of errors th](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally)

Also we are console logging the error, but if that happens the user has no idea what's going on.
I suggest adding in a error state and displaying that if anything goes wrong.